server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java

```diff
@@ -0,0 +1,376 @@
+
+    protected abstract void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommit,
+                                            ActionListener<TransportResponse.Empty> responseActionListener);
+
+    @Override
+    public String toString() {
+        return "Publication{term=" + publishRequest.getAcceptedState().term() +
+            ", version=" + publishRequest.getAcceptedState().version() + '}';
+    }
+
+    enum PublicationTargetState {
+        NOT_STARTED,
+        FAILED,
+        SENT_PUBLISH_REQUEST,
+        WAITING_FOR_QUORUM,
+        SENT_APPLY_COMMIT,
+        APPLIED_COMMIT,
+    }
+
+    static class PublicationTargetStateMachine {
+        private PublicationTargetState state = PublicationTargetState.NOT_STARTED;
+
+        public void setState(PublicationTargetState newState) {
+            switch (newState) {
+                case NOT_STARTED:
+                    assert false : state + " -> " + newState;
+                    break;
+                case SENT_PUBLISH_REQUEST:
+                    assert state == PublicationTargetState.NOT_STARTED : state + " -> " + newState;
+                    break;
+                case WAITING_FOR_QUORUM:
+                    assert state == PublicationTargetState.SENT_PUBLISH_REQUEST : state + " -> " + newState;
+                    break;
+                case SENT_APPLY_COMMIT:
+                    assert state == PublicationTargetState.WAITING_FOR_QUORUM : state + " -> " + newState;
+                    break;
+                case APPLIED_COMMIT:
+                    assert state == PublicationTargetState.SENT_APPLY_COMMIT : state + " -> " + newState;
+                    break;
+                case FAILED:
+                    assert state != PublicationTargetState.APPLIED_COMMIT : state + " -> " + newState;
+                    break;
+            }
+            state = newState;
+        }
+
+        public boolean isActive() {
+            return state != PublicationTargetState.FAILED
+                && state != PublicationTargetState.APPLIED_COMMIT;
+        }
+
+        public boolean isWaitingForQuorum() {
+            return state == PublicationTargetState.WAITING_FOR_QUORUM;
+        }
+
+        public boolean mayCommitInFuture() {
+            return (state == PublicationTargetState.NOT_STARTED
+                || state == PublicationTargetState.SENT_PUBLISH_REQUEST
+                || state == PublicationTargetState.WAITING_FOR_QUORUM);
+        }
+
+        public boolean isFailed() {
+            return state == PublicationTargetState.FAILED;
+        }
+
+        @Override
+        public String toString() {
+            return state.toString();
+        }
+    }
+
+    private class PublicationTarget {
+        private final DiscoveryNode discoveryNode;
+        private final PublicationTargetStateMachine publicationTargetStateMachine = new PublicationTargetStateMachine();
```

Comment:
This was useful for development, but now things are more stable shall we merge this into `PublicationTarget`?

Reply:
ok, fixed in b7b4e39b980